Reviewing the algorithm for calculating workshop balances

viktorcsimma opened the pull request:
There was a strange phenomenon with the workshop balances: their sum was more than 49% of the amount received from kkt payments; in theory, it should be between 30% and 40%.

I've rewritten the algorithm calculating the data in the balances table; now it displays less people paying kkt and it seems to be consistent with the income numbers. @kdmnk could you check it out?

Commit:
Rewritten the algorithm for calculating workshop balances; seems to be consistent with the number of transactions now

kdmnk commented:
I see, yet I still do not understand where the difference is coming from.
You changed two things:
- Check if the user paid: Instead of checking if the user is not returned by the hasToPayKKTNetregInSemester scope (which returns all active collegists who don't have either kkt or netreg transaction in the semester), you check if the individual user has any kkt transaction (which is more database calls, but that's another problem). Since active memberships are checked anyway before, this change should not result in any difference (given that everyone who has netreg transaction also has kkt transaction, and the contrary). 
- Use the paid amount instead of the amount in the config. I see the point, but do we want to have a workshop where they get less amount only because they have some people in the student council who pay 0 ft?  My opinion is that the workshop balance should be calculated based solely on the number of members, and not who paid how much. 

How did you check and validate the new numbers?

If we want to fix this in the long term, write test cases. This module is one of the last crucial parts where we don't have tests yet.
Let me know what you think.

Commit:
Changing kkt amount used for workshop balances back to the content of the config file

Commit:
Adding documentation comments

viktorcsimma commented:
For the amount part, I agree with you; I've changed it back now. For the other part, I'm thinking about how to solve that. Now I've left it, and it seems to work; but if I understand well, the problem with those paying 0 Ft persists. (Or might that be the reason for the difference?)

Commit:
Changing payedKKT to be nullable; checking based on that

viktorcsimma commented:
Now I've changed it to nullable; the numbers are still OK, but if I understand right, there aren't any transactions recorded for student council members, so the problem would still persist. Is that correct?

I've actually tested the code on the live site; it was consistent with the real values, but I still don't know the reason. Maybe I'll take a look at `hasToPayKKTNetregInSemester`.

kdmnk commented:
I insist on creating tests for this. There are multiple examples now in the project. That should also reveal the answer to our questions.

viktorcsimma committed:
Adding a test; fails for both versions

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 9c78cff according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/342

viktorcsimma committed:
Simplifying test

viktorcsimma commented:
Have you thought of something like this? I would write more, of course.

Actually, after some seeding, this fails for both the old and the new version failed... There, the sum of workshop balances was 10200 in the old version and 6600 in the new version; while the income from kkt was always 24000. So both of them was too little. I will check this out.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,49 @@
+     *
+     * @return void
+     */
+    public function testWorkshopRatio()
+    {
+        // Should we generate users here?
+
+        $transactions = \App\Models\Transaction::whereIn(
+            'payment_type_id',
+            [\App\Models\PaymentType::kkt()->id]
+        )
+            ->where('semester_id', \App\Models\Semester::current()->id)
+            ->get();
+        $skkt = $transactions->where('payment_type_id', \App\Models\PaymentType::kkt()->id)
+            ->sum('amount');
+
+        \App\Models\WorkshopBalance::generateBalances(\App\Models\Semester::current()->id);
+        $sum = \App\Models\WorkshopBalance::all()->sum('allocated_balance');
+
+        /*
+        echo($skkt);
+        echo("\n");
+        echo($sum);
+        echo("\n");
+        echo(config('custom.workshop_balance_resident'));
+        echo("\n");
+        echo(config('custom.workshop_balance_extern'));
+        echo("\n");
+        */
+
+        $diff1 = config('custom.workshop_balance_resident') * $skkt - $sum;
+        $diff2 = config('custom.workshop_balance_extern') * $skkt - $sum;
+        $this->assertTrue($diff1 * $diff2 <= 0);
+    }
```
Aren't all the values 0 here?

You should test if after someone pays, the workshop balance for the user got modified with the correct amount.
Test externs, residents, with single and multiple workshops.

You can keep this test case but this is not so useful on its own.

viktorcsimma committed:
Further correcting test; still fails

viktorcsimma committed:
Adding new test; restructuring code for kkt payments

viktorcsimma committed:
Fixing normal kkt payment process

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 7f47805 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/342

viktorcsimma commented:
Now I think I've written something like that:) @kdmnk could you check it out?

Maybe a feature test for a complete HTTP request would also be useful; but that could be done in a separate PR, if you agree.

viktorcsimma commented:
One more thing: I had to rewrite `isResident()` and `isExtern()` because I got an error that there is a string somewhere where a model object should be. Do you have an idea what could have caused this?

viktorcsimma commented:
And maybe there could be another case where the test user has multiple workshops.

kdmnk commented on app/Http/Controllers/StudentsCouncil/EconomicController.php:
```diff
@@ -85,41 +85,24 @@ public function payKKTNetreg(Request $request)
         $validator->validate();
 
         $payer = User::findOrFail($request->user_id);
+        $return_values = $payer->payKKTNetreg($request->kkt, $request->netreg);
```
[$kkt, $netreg, $new_internet_expire_date] = $payer->payKKTNetreg(...)

kdmnk commented on app/Models/User.php:
```diff
@@ -681,13 +681,10 @@ public function setCollegist($objectName): void
-            ->exists();
-        }
-        return $this->hasRole([Role::COLLEGIST => Role::RESIDENT]);
+        return $this->roles()
+        ->where('role_id', Role::collegist()->id)
+        ->where('object_id', RoleObject::firstWhere('name', Role::RESIDENT)->id)
+        ->exists();
```
Just comment the previous one out and add a todo. 
We need to remove the first branch when we detach roles from the applicants so it should just be  $this->hasRole(...).

For the bug you see, idk, we should find it asap, but if you can not you can leave this like that.

viktorcsimma replied:
Oh; it was the cache:D After running `php artisan optimize:clear`, it was fine. I revert it.

kdmnk commented on app/Models/User.php:
```diff
@@ -869,25 +863,64 @@ public function sumOfActiveFreePages(): int
+        // Creating transactions
+        $kkt = Transaction::create([
+            'checkout_id' => Checkout::studentsCouncil()->id,
+            'receiver_id' => Checkout::studentsCouncil()->handler_id,
```
the receiver can be any collegist that the student council delegates to, not necessarily the handler. In both cases.

kdmnk replied:
Despite how they do this nowadays.

viktorcsimma replied:
I now added it as a parameter.

kdmnk commented on app/Models/User.php:
```diff
@@ -869,25 +863,64 @@ public function sumOfActiveFreePages(): int
+     * @return ?int
      */
-    public function payedKKTInSemester(Semester $semester): int
+    public function payedKKTInSemester(Semester $semester): ?int
```
Would you rename everything to "paidSomething" while you are here, please?

viktorcsimma replied:
I don't understand this here; should I generalise the function to be able to pay anything? Now it is specially about kkt.

kdmnk replied:
Just rename it to make it gramatically correct, paidKKT and paidNetreg.

viktorcsimma replied:
Oh, I see:)

kdmnk commented on app/Models/Workshop.php:
```diff
@@ -69,6 +70,23 @@ public function externs()
+     */
+    public function balance(int $semester = null): WorkshopBalance
+    {
+        return $this->balances()->where('semester_id', $semester ?? Semester::current()->id)->firstOrFail();
```
Won't this die if we switch to next semester and the current workshop balances are not generated yet?
I'd allow null value and handle that.

kdmnk commented on app/Models/WorkshopBalance.php:
```diff
@@ -58,7 +65,7 @@ public static function generateBalances($semester_id)
             foreach ($workshop->users as $member) {
                 if (isset($active_users[$member->id])) {
-                    if (!isset($users_has_to_pay_kktnetreg[$member->id])) {
+                    if (!is_null($member->payedKKTInSemester(Semester::find($semester_id)))) {
```
Get the semester outside of the loop. Or even outside this function and get the semester model as param. It makes our type checking stronger.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,104 @@
+     */
+    public function testPayment()
+    {
+        foreach ([true, false] as $is_extern) {
```
nice trick, but maybe cleaner to put these into two separate test cases.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,104 @@
+    {
+        foreach ([true, false] as $is_extern) {
+            $workshop_id = rand(1, count(Workshop::ALL));
+            $workshop = Workshop::findOrFail($workshop_id);
```
You can use Workshop::all()->random() and then get the id from that.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,104 @@
+            if ($is_extern) {
+                $user->setExtern();
+            } else {
+                $user->setCollegist(\App\Models\Role::RESIDENT);
```
->setResident?

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,104 @@
+
+            $user->setStatusFor(Semester::current(), SemesterStatus::ACTIVE);  // important!
+
+            $user->workshops()->attach($workshop_id);
```
FYI I believe you can simply pass the workshop, but whatever

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,104 @@
+
+            $user->workshops()->attach($workshop_id);
+
+            $with_same_status =
+                $is_extern ? $workshop->externs() : $workshop->residents();
+            $this->assertTrue(
+                $with_same_status->filter(function (User $u) use ($user) {return $u->id == $user->id;})->count()
+                  == 1
+            );
```
What's the point here? this only checks that you set previously and that the user does not appear twice in the collection which can never be the case, can it?

viktorcsimma replied:
The point was that I wanted to know that the status setting occurred successfully. But I agree it does not belong here.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,104 @@
+            $balance = $workshop->balance();
+
+            $old = $balance->allocated_balance;
+            $user->payKKTNetreg(config('custom.kkt'), config('custom.netreg'));
+            $balance->refresh();
+            $new = $balance->allocated_balance;
+
+            $this->assertTrue($user->payedKKT() == config('custom.kkt'));
+
+            $this->assertTrue($new - $old == round(
+                ($is_extern ? config('custom.workshop_balance_extern')
+                                                           : config('custom.workshop_balance_resident'))
+                                               * config('custom.kkt')
```
Use precoded values instead of config values to make this easier to read and debug. You can have another test case that the pay function uses the config values accordingly.

viktorcsimma replied:
I had a problem with the ratios, or rather a design question: now `generateBalances` itself uses the config values for them; should we provide them as parameters or store them as properties of the semester? Or simply leave this as it is?

kdmnk reviewed (changes requested):
Okay looking better. Please take a look at my comments. And yes, please write a case to test multiple workshops, that's a common case we have. So I imagine 4 cases, combining resident/extern and one/multiple workshops. 
Have you found the original bug yet?

An HTTP test is not necessarily here as we test those anyways during development. Those are useful for refactorings though, but we most importantly lack unit tests to test different edge cases like these.

viktorcsimma committed:
Addressing most of the requests

viktorcsimma committed:
Reverting changes to isResident and isExtern

viktorcsimma committed:
Simplifying code, using constants for test values (but not for ratios)

viktorcsimma committed:
Adding new test cases for users with multiple workshops

viktorcsimma committed:
Removing the first test case (it did not work properly)

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 0db132a according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/342

viktorcsimma commented:
Now I've added new test cases; I've also removed the first one (there was some problem with it, and you have mentioned it's not useful anyway).

Can you check out these and my questions, too? Sometime when it's convenient for you – until then, happy holidays;)

kdmnk commented on app/Http/Controllers/StudentsCouncil/EconomicController.php:
```diff
@@ -85,41 +85,22 @@ public function payKKTNetreg(Request $request)
 
         $payer = User::findOrFail($request->user_id);
+        // the current user will be the receiver
+        [$kkt, $netreg, $new_internet_expire_date] = $payer->payKKTNetreg(Auth::user()->id, $request->kkt, $request->netreg);
```
As a new thought, this adds another method to the User model which we want to keep as simple as possible.
I was thinking a lot about where would this be placed the best, in the Transaction model or somehow in CheckoutHandler, but these actions are typical controller actions, so you don't need to extract them away. We can make the function smaller with automatic email sending with observers, but that would mean refactoring the whole CheckoutHandler.
Just revert this change for now, sorry :D

viktorcsimma replied:
What about writing a static method in the EconomicController class? That was where this code was originally written. I would rather write a method for it somewhere, so that we don't need to duplicate the code in the test. See the code I'm going to push soon.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+            $user->workshops()->attach($workshop);
+        }
+
+        $this->assertTrue(is_null($user->payedKKT()));
```
unnecessary

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+        $this->assertTrue(is_null($user->payedKKT()));
+        WorkshopBalance::generateBalances(Semester::current()); // this also ensures the balance won't be null
+
+        $balances = array_map(function ($workshop) {return $workshop->balance();}, $workshops);
```
use collection methods. 
$old_balances = $workshops->map(fn ($workshop) => $workshop->balance()->allocated_balance). 
$new_balances = same as above, i don't even think you need to refresh the vars this way.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+        $news = array_map(function ($balance) {return $balance->fresh()->allocated_balance;}, $balances);
+
+        $amount_for_workshops =
+            ($is_extern ? config('custom.workshop_balance_extern')
```
remove these configs as well. This also ensures that the pipelines test with the same data.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+            ($is_extern ? config('custom.workshop_balance_extern')
+                        : config('custom.workshop_balance_resident'))
+                * self::TEST_KKT;
+        for ($i = 0; $i < $number_of_workshops; ++$i) {
+            // to avoid floating point equality checking
+            // and to allow some error from rounding:
+            $this->assertTrue(abs(
+                $number_of_workshops * ($news[$i] - $olds[$i])
+                                - $amount_for_workshops
+            ) < self::THRESHOLD_DIFF);
+        }
```
This is not readable. Use assertEqualsWithDelta and avoid for-loop indexes. Loop through the workshops, use dicts or whatever. Even easier to check if the workshop balances equal to an expected hard-coded value that you calculated before, no need to calculate the diffs programatically. Trust me it's better for debugging tests.

You should be able to set the threshold below 0.1. If not, something's wrong. If we round, still, the sum should all add up to the originally paid amount. Maybe you need to modify the logic behind it.

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+            $user->setResident();
+        }
+
+        $user->setStatusFor(Semester::current(), SemesterStatus::ACTIVE);  // important!
```
->setStatus(SemesterStatus::ACTIVE);

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        $user->setStatusFor(Semester::current(), SemesterStatus::ACTIVE);  // important!
+        foreach($workshops as $workshop) {
+            $user->workshops()->attach($workshop);
+        }
```
$user->workshops()->sync(Workshop::all()->random(2))

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+        $workshops = Workshop::all()->random($number_of_workshops)->all();
+        $user = User::factory()->create();
+
+        if ($is_extern) {
+            $user->setExtern();
+        } else {
+            $user->setResident();
+        }
```
$user->setCollegist($is_extern ? Role::EXTERN : Role::RESIDENT)

kdmnk commented on tests/Unit/EconomicTest.php:
```diff
@@ -0,0 +1,114 @@
+     */
+    public function testExternPaymentWithTwoWorkshops()
+    {
+        $this->testPayment(true, 2);
```
These are also not readable. How about passing the user as a parameter?
I suggested some changes to make the user creation logic smaller above. 

I might also move the assertion logic here with the hard-coded amounts described above.
And with that, you probably won't even need these functions :D

viktorcsimma replied:
Passing the user is a good idea; that way, it's more evident what we are doing while we also avoid much of code duplication. I think I'll go for that:)

kdmnk reviewed (changes requested):
These comments are just to make the code more readable. Good job

kdmnk commented:
I'm not saying that it's perfect, but you can take a look at these tests, here the assertions are clear what they tests and how they got calculated (eg. hard coded  `400 * (100 - (100 - 25) / 2)`):
https://github.com/DevsWithDodo/dodo_api/blob/master/tests/Feature/PurchaseCreateTest.php

viktorcsimma committed:
Applying suggestions

viktorcsimma committed:
Restructuring payment logic into EconomicController::payKKTNetregLogic

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in b24ede1 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/342

kdmnk commented:
Sorry about the hassle of such a little feature, but I think it looks really good and clean now. Merge as you wish

viktorcsimma commented:
It's fine; thank you for your help:) eventually, I'll have to know how to sustain this code quality.
